# 01 - Data Cleaning & Preprocessing
## SF EMS Emergency Response Dataset

**Period**: Sep 2025 – Aug 2026  
**Owner**: P2 (Data Engineer) + P1 (Team Lead)  
**Output**: `sf_ems_clean_9_2025_to_8_2026.csv`

---
## 1. Load Data

In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.width', 300)
pd.set_option('display.float_format', '{:.2f}'.format)

In [2]:
# INPUT = raw SF EMS data MERGED with web-scraped weather & holidays
# (sf_ems_weather_merged.csv is produced by 03_web_scraping.ipynb)
RAW_PATH = r'C:\Users\LOQ\OneDrive\Desktop\Emergancy\sf-ems-project\notebooks\sf_ems_weather_merged.csv'
OUTPUT_PATH = r'C:\Users\LOQ\OneDrive\Desktop\Emergancy\sf-ems-project\data\cleaned\sf_ems_clean_9_2025_to_8_2026.csv'

df = pd.read_csv(RAW_PATH, low_memory=False)
print(f'Loaded: {df.shape[0]:,} rows x {df.shape[1]} columns')

Loaded: 359,542 rows x 44 columns


---
## 2. Initial Inspection

In [3]:
df.head()

,call_number,unit_id,incident_number,call_type,call_date,watch_date,received_dttm,entry_dttm,dispatch_dttm,response_dttm,on_scene_dttm,transport_dttm,hospital_dttm,call_final_disposition,available_dttm,address,city,zipcode_of_incident,battalion,station_area,box,original_priority,priority,final_priority,als_unit,call_type_group,number_of_alarms,unit_type,unit_sequence_in_call_dispatch,fire_prevention_district,supervisor_district,neighborhoods_analysis_boundaries,rowid,case_location,data_as_of,data_loaded_at,date,avg_temperature,min_temperature,max_temperature,precipitation,max_wind_speed,holiday_name,is_holiday
0,252440002,AM118,25124149,Medical Incident,2025-09-01T00:00:00.000,2025-08-31T00:00:00.000,2025-09-01 00:00:46,2025-09-01T00:00:46.000,2025-09-01T00:01:32.000,2025-09-01T00:01:51.000,NaN,NaN,NaN,Code 2 Transport,2025-09-01T00:03:00.000,GEARY ST/HYDE ST,San Francisco,94109.00,B04,3.00,1556,A,2,2,False,Non Life-threatening,1,PRIVATE,1,4.00,3.00,Tenderloin,252440002-AM118,POINT (-122.416449123 37.786310218),2025-09-01T03:32:33.000,2025-09-02T04:13:45.000,2025-09-01,19.80,15.40,25.40,0.00,15.30,Labor Day,True
1,252440009,M562,25124151,Medical Incident,2025-09-01T00:00:00.000,2025-08-31T00:00:00.000,2025-09-01 00:01:03,2025-09-01T00:05:46.000,2025-09-01T00:08:14.000,2025-09-01T00:08:20.000,2025-09-01T00:29:39.000,NaN,NaN,Other,2025-09-01T01:19:59.000,BEMIS ST/MOFFITT ST/ROANOKE ST,San Francisco,94131.00,B06,26.00,8111,2,2,2,True,Non Life-threatening,1,MEDIC,1,6.00,8.00,Glen Park,252440009-M562,POINT (-122.431348921 37.736617894),2025-09-01T03:32:33.000,2025-09-02T04:13:45.000,2025-09-01,19.80,15.40,25.40,0.00,15.30,Labor Day,True
2,252440008,CP5,25124150,Medical Incident,2025-09-01T00:00:00.000,2025-08-31T00:00:00.000,2025-09-01 00:02:22,2025-09-01T00:05:26.000,2025-09-01T00:11:36.000,2025-09-01T00:11:36.000,2025-09-01T00:37:05.000,NaN,NaN,Other,2025-09-01T01:07:12.000,FULTON ST/STANYAN ST,San Francisco,94122.00,B07,21.00,4556,2,2,2,False,Non Life-threatening,1,CP,1,7.00,1.00,Lone Mountain/USF,252440008-CP5,POINT (-122.454682648 37.774755484),2025-09-01T03:32:33.000,2025-09-02T04:13:45.000,2025-09-01,19.80,15.40,25.40,0.00,15.30,Labor Day,True
3,252440008,SCRT10,25124150,Medical Incident,2025-09-01T00:00:00.000,2025-08-31T00:00:00.000,2025-09-01 00:02:22,2025-09-01T00:05:26.000,2025-09-01T00:06:59.000,2025-09-01T00:07:03.000,NaN,NaN,NaN,Other,2025-09-01T00:11:46.000,FULTON ST/STANYAN ST,San Francisco,94122.00,B07,21.00,4556,2,2,2,False,Non Life-threatening,1,CP,2,7.00,1.00,Lone Mountain/USF,252440008-SCRT10,POINT (-122.454682648 37.774755484),2025-09-01T03:32:33.000,2025-09-02T04:13:45.000,2025-09-01,19.80,15.40,25.40,0.00,15.30,Labor Day,True
4,252440014,KM101,25124152,Medical Incident,2025-09-01T00:00:00.000,2025-08-31T00:00:00.000,2025-09-01 00:06:00,2025-09-01T00:07:40.000,2025-09-01T00:08:37.000,2025-09-01T00:11:53.000,2025-09-01T00:11:53.000,2025-09-01T00:46:18.000,2025-09-01T00:51:00.000,Code 3 Transport,2025-09-01T01:34:49.000,STEINER ST/WALLER ST,San Francisco,94117.00,B05,6.00,3633,2,2,2,False,Non Life-threatening,1,PRIVATE,1,5.00,5.00,Haight Ashbury,252440014-KM101,POINT (-122.431904324 37.770964426),2025-09-01T03:32:33.000,2025-09-02T04:13:45.000,2025-09-01,19.80,15.40,25.40,0.00,15.30,Labor Day,True


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359542 entries, 0 to 359541
Data columns (total 44 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   call_number                        359542 non-null  int64  
 1   unit_id                            359542 non-null  object 
 2   incident_number                    359542 non-null  int64  
 3   call_type                          359542 non-null  object 
 4   call_date                          359542 non-null  object 
 5   watch_date                         359542 non-null  object 
 6   received_dttm                      359542 non-null  object 
 7   entry_dttm                         359542 non-null  object 
 8   dispatch_dttm                      359542 non-null  object 
 9   response_dttm                      349916 non-null  object 
 10  on_scene_dttm                      281553 non-null  object 
 11  transport_dttm                     8641

In [5]:
df.describe()

,call_number,incident_number,zipcode_of_incident,station_area,final_priority,number_of_alarms,unit_sequence_in_call_dispatch,fire_prevention_district,supervisor_district,avg_temperature,min_temperature,max_temperature,precipitation,max_wind_speed
count,359542.00,359542.00,359087.00,359531.00,359542.00,359542.00,359542.00,357400.00,359474.00,359542.00,359542.00,359542.00,359542.00,359542.00
mean,258352781.04,25745553.78,94113.25,17.64,2.73,1.00,1.97,4.58,5.84,14.37,11.63,18.16,1.70,17.80
std,3918191.51,431844.52,10.85,14.43,0.45,0.05,1.77,2.86,2.70,2.85,2.76,3.93,5.22,5.11
min,252440002.00,25124149.00,94102.00,1.00,2.00,1.00,1.00,1.00,1.00,7.00,4.40,10.00,0.00,6.10
25%,253322078.00,25169515.00,94103.00,4.00,2.00,1.00,1.00,2.00,4.00,13.00,10.30,15.70,0.00,14.10
50%,260540710.50,26027677.50,94110.00,13.00,3.00,1.00,1.00,4.00,6.00,14.70,12.00,17.80,0.00,17.60
75%,261452238.25,26074274.75,94121.00,32.00,3.00,1.00,2.00,7.00,8.00,16.30,13.60,20.10,0.10,20.70
max,262384064.00,26120908.00,94158.00,94.00,3.00,3.00,61.00,10.00,11.00,22.40,18.20,31.50,43.00,42.70


### 2.1 Check Date Range

In [6]:
print('Call Date range:', pd.to_datetime(df['call_date'], errors='coerce').min(), 'to', pd.to_datetime(df['call_date'], errors='coerce').max())
print('Received DtTm range:', pd.to_datetime(df['received_dttm'], errors='coerce').min(), 'to', pd.to_datetime(df['received_dttm'], errors='coerce').max())

Call Date range: 2025-09-01 00:00:00 to 2026-08-26 00:00:00
Received DtTm range: 2025-09-01 00:00:46 to 2026-08-26 23:56:48


---
## 3. Column Cleanup

In [7]:
print('Original columns:')
for i, col in enumerate(df.columns, 1):
    print(f'  {i:2d}. {col}')

Original columns:
   1. call_number
   2. unit_id
   3. incident_number
   4. call_type
   5. call_date
   6. watch_date
   7. received_dttm
   8. entry_dttm
   9. dispatch_dttm
  10. response_dttm
  11. on_scene_dttm
  12. transport_dttm
  13. hospital_dttm
  14. call_final_disposition
  15. available_dttm
  16. address
  17. city
  18. zipcode_of_incident
  19. battalion
  20. station_area
  21. box
  22. original_priority
  23. priority
  24. final_priority
  25. als_unit
  26. call_type_group
  27. number_of_alarms
  28. unit_type
  29. unit_sequence_in_call_dispatch
  30. fire_prevention_district
  31. supervisor_district
  32. neighborhoods_analysis_boundaries
  33. rowid
  34. case_location
  35. data_as_of
  36. data_loaded_at
  37. date
  38. avg_temperature
  39. min_temperature
  40. max_temperature
  41. precipitation
  42. max_wind_speed
  43. holiday_name
  44. is_holiday


In [8]:
df.columns = df.columns.str.strip().str.lower()
df.columns = df.columns.str.replace(r'\s+', '_', regex=True)
df.columns = df.columns.str.replace(r'[^a-z0-9_]', '_', regex=True)
df.columns = df.columns.str.replace(r'_+', '_', regex=True)
df.columns = df.columns.str.strip('_')

print('Cleaned columns:')
for i, col in enumerate(df.columns, 1):
    print(f'  {i:2d}. {col}')

Cleaned columns:
   1. call_number
   2. unit_id
   3. incident_number
   4. call_type
   5. call_date
   6. watch_date
   7. received_dttm
   8. entry_dttm
   9. dispatch_dttm
  10. response_dttm
  11. on_scene_dttm
  12. transport_dttm
  13. hospital_dttm
  14. call_final_disposition
  15. available_dttm
  16. address
  17. city
  18. zipcode_of_incident
  19. battalion
  20. station_area
  21. box
  22. original_priority
  23. priority
  24. final_priority
  25. als_unit
  26. call_type_group
  27. number_of_alarms
  28. unit_type
  29. unit_sequence_in_call_dispatch
  30. fire_prevention_district
  31. supervisor_district
  32. neighborhoods_analysis_boundaries
  33. rowid
  34. case_location
  35. data_as_of
  36. data_loaded_at
  37. date
  38. avg_temperature
  39. min_temperature
  40. max_temperature
  41. precipitation
  42. max_wind_speed
  43. holiday_name
  44. is_holiday


---
## 4. Type Conversion — Timestamps

All 7 time-phase columns are strings. Convert them to datetime.

In [9]:
timestamp_cols = [
    'received_dttm', 'entry_dttm', 'dispatch_dttm',
    'response_dttm', 'on_scene_dttm', 'transport_dttm',
    'hospital_dttm', 'available_dttm'
]

for col in timestamp_cols:
    df[col] = pd.to_datetime(df[col], errors='coerce')
    print(f'{col}: {df[col].dtype}')

received_dttm: datetime64[ns]
entry_dttm: datetime64[ns]
dispatch_dttm: datetime64[ns]
response_dttm: datetime64[ns]
on_scene_dttm: datetime64[ns]
transport_dttm: datetime64[ns]
hospital_dttm: datetime64[ns]
available_dttm: datetime64[ns]


---
## 5. Null Analysis & Handling

In [10]:
nulls = df.isnull().sum()
null_pct = (nulls / len(df)) * 100

null_report = pd.DataFrame({
    'null_count': nulls,
    'null_pct': null_pct
}).query('null_count > 0').sort_values('null_pct', ascending=False)

print('Columns with nulls:')
print(null_report.to_string())
print(f'\nTotal rows: {len(df):,}')

Columns with nulls:
                                   null_count  null_pct
holiday_name                           346103     96.26
hospital_dttm                          274560     76.36
transport_dttm                         273132     75.97
on_scene_dttm                           77989     21.69
response_dttm                            9626      2.68
call_type_group                          7953      2.21
original_priority                        3248      0.90
fire_prevention_district                 2142      0.60
zipcode_of_incident                       455      0.13
city                                      363      0.10
neighborhoods_analysis_boundaries          73      0.02
supervisor_district                        68      0.02
address                                    19      0.01
case_location                              15      0.00
station_area                               11      0.00
box                                        10      0.00
available_dttm              

### 5.1 Null Handling Strategy

Three kinds of nulls:

1. **Truly missing** -> imputed
2. **Not applicable** -> kept as NaN (a value factually does not exist)
3. **Unknown outcome** -> kept as NaN (cannot be invented)

| Column | Null % | Kind | Handling |
|--------|--------|------|----------|
| `transport_dttm` / `hospital_dttm` / `transport_time_min` | ~76% | Not applicable (no patient transport) | Keep NaN |
| `on_scene_dttm` / `response_time_min` | 21.7% | Unknown outcome (cancelled / no on-scene) | Keep NaN; ML drops these only when predicting response time |
| `response_dttm` | 2.7% | Missing | Keep NaN |
| `holiday_name` | 96% | Not applicable (not a holiday) | Keep NaN; use `is_holiday` flag |
| `call_type_group` | 2.2% | Missing | Fill `'Unknown'` |
| `original_priority` | 0.9% | Missing | Fill from `priority` |
| `fire_prevention_district` | 0.6% | Missing | Fill mode |
| `zipcode_of_incident` | 0.1% | Missing | Fill mode |
| `city` | 0.1% | Missing | Fill `'San Francisco'` |
| `neighborhoods_analysis_boundaries` | ~0% | Missing | Fill `'Unknown'` |
| `supervisor_district` | ~0% | Missing | Fill mode |
| `station_area` / `box` | ~0% | Missing | Fill mode |

**Why NOT impute the time targets?**
- `response_time_min` is the ML outcome: imputing it would fabricate results the model would then "predict". Rows without an on-scene time are dropped only inside the ML step.
- `transport_time_min` cleanly does not exist for the ~76% of calls without a patient transport.

In [11]:
# --- Impute only the truly-missing values ---
df['call_type_group'] = df['call_type_group'].fillna('Unknown')
df['neighborhoods_analysis_boundaries'] = df['neighborhoods_analysis_boundaries'].fillna('Unknown')

df['original_priority'] = df['original_priority'].fillna(df['priority'])
df['city'] = df['city'].fillna('San Francisco')

for col in ['zipcode_of_incident', 'fire_prevention_district', 'supervisor_district', 'station_area', 'box']:
    mode_val = df[col].mode(dropna=True)
    if not mode_val.empty:
        df[col] = df[col].fillna(mode_val.iloc[0])

print('Remaining nulls after imputation:')
print(df[['city', 'original_priority', 'call_type_group', 'zipcode_of_incident',
          'neighborhoods_analysis_boundaries', 'fire_prevention_district',
          'supervisor_district', 'station_area', 'box']].isnull().sum())
print('\nStill-NaN (kept on purpose):')
still = ['response_dttm', 'on_scene_dttm', 'transport_dttm', 'hospital_dttm', 'holiday_name']
print(df[still].isnull().sum())

Remaining nulls after imputation:
city                                 0
original_priority                    0
call_type_group                      0
zipcode_of_incident                  0
neighborhoods_analysis_boundaries    0
fire_prevention_district             0
supervisor_district                  0
station_area                         0
box                                  0
dtype: int64

Still-NaN (kept on purpose):
response_dttm       9626
on_scene_dttm      77989
transport_dttm    273132
hospital_dttm     274560
holiday_name      346103
dtype: int64


---
## 6. Duplicate Check

In [12]:
dup_count = df.duplicated(subset=['rowid']).sum()
print(f'Duplicate RowID count: {dup_count}')

if dup_count > 0:
    print('\nDuplicate rows:')
    print(df[df.duplicated(subset=['rowid'], keep=False)].sort_values('rowid').head(10))
else:
    print('No duplicates found.')

Duplicate RowID count: 0
No duplicates found.


---
## 7. Feature Engineering

Create derived columns needed for EDA and modeling.

In [13]:
df['response_time_min'] = (df['on_scene_dttm'] - df['received_dttm']).dt.total_seconds() / 60
df['dispatch_lag_min'] = (df['dispatch_dttm'] - df['entry_dttm']).dt.total_seconds() / 60
df['transport_time_min'] = (df['hospital_dttm'] - df['transport_dttm']).dt.total_seconds() / 60

print('response_time_min:')
print(f'  Non-null: {df["response_time_min"].notna().sum():,}')
print(f'  Null: {df["response_time_min"].isna().sum():,}')
print(f'  Mean: {df["response_time_min"].mean():.2f} min')
print(f'  Median: {df["response_time_min"].median():.2f} min')

print('\ndispatch_lag_min:')
print(f'  Non-null: {df["dispatch_lag_min"].notna().sum():,}')
print(f'  Mean: {df["dispatch_lag_min"].mean():.2f} min')

response_time_min:
  Non-null: 281,553
  Null: 77,989
  Mean: 11.45 min
  Median: 8.22 min

dispatch_lag_min:
  Non-null: 359,542
  Mean: 2.05 min


In [14]:
df['hour_of_day'] = df['received_dttm'].dt.hour
df['day_of_week'] = df['received_dttm'].dt.day_name()
df['month'] = df['received_dttm'].dt.month
df['year'] = df['received_dttm'].dt.year
df['is_weekend'] = df['day_of_week'].isin(['Saturday', 'Sunday'])
df['is_als'] = df['als_unit'].astype(int) #ALS = Advanced Life Support 

print('New time columns created:')
print(df[['received_dttm', 'hour_of_day', 'day_of_week', 'month', 'is_weekend', 'is_als']].head())

New time columns created:
        received_dttm  hour_of_day day_of_week  month  is_weekend  is_als
0 2025-09-01 00:00:46            0      Monday      9       False       0
1 2025-09-01 00:01:03            0      Monday      9       False       1
2 2025-09-01 00:02:22            0      Monday      9       False       0
3 2025-09-01 00:02:22            0      Monday      9       False       0
4 2025-09-01 00:06:00            0      Monday      9       False       0


---
## 7b. IQR Outlier Detection (Flag-Only)

Standard Tukey fences (1.5x IQR) on the derived time columns.
- **Flag**: `is_outlier_*` columns added (row = True if outside the fence)
- **Values are NOT changed** - the CSV keeps the real measured times
- **Negative times** (data inconsistencies) fall below the lower fence (floored at 0) -> automatically flagged

Why flag-only instead of capping?
- Capping **crushes model performance**: `R2 = 0.56` on capped response times vs `R2 = 0.80` on the real values.
- Keep the real values; filter with the flags in Excel / Power BI / ML when desired (`is_outlier_* == False`).

In [15]:
iqr_cols = ['response_time_min', 'dispatch_lag_min', 'transport_time_min']

for col in iqr_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = max(0, Q1 - 1.5 * IQR)
    upper = Q3 + 1.5 * IQR

    is_outlier = (df[col] < lower) | (df[col] > upper)
    df[f'is_outlier_{col}'] = is_outlier

    outlier_count = is_outlier.sum()
    pct = outlier_count / len(df) * 100
    print(f'{col}:')
    print(f'  Q1={Q1:.2f}, Q3={Q3:.2f}, IQR={IQR:.2f}')
    print(f'  Fences: [{lower:.2f}, {upper:.2f}]')
    print(f'  Outliers flagged: {outlier_count:,} ({pct:.2f}%)')
    print(f'  Values unchanged: min={df[col].min():.2f}, max={df[col].max():.2f}')

print('\nFlag-only complete: is_outlier_* columns added, values NOT modified.')

response_time_min:
  Q1=6.07, Q3=13.15, IQR=7.08
  Fences: [0.00, 23.77]
  Outliers flagged: 22,094 (6.15%)
  Values unchanged: min=-101.10, max=3609.13
dispatch_lag_min:
  Q1=0.27, Q3=1.10, IQR=0.83
  Fences: [0.00, 2.35]
  Outliers flagged: 41,976 (11.67%)
  Values unchanged: min=-59.17, max=3609.13
transport_time_min:
  Q1=9.92, Q3=23.33, IQR=13.42
  Fences: [0.00, 43.46]
  Outliers flagged: 2,522 (0.70%)
  Values unchanged: min=-712.82, max=140.60

Flag-only complete: is_outlier_* columns added, values NOT modified.


---
## 8. Date Filtering (Verify Sep 2025 – Aug 2026)

In [16]:
print('Date range before filtering:')
print(f'  Min: {df["received_dttm"].min()}')
print(f'  Max: {df["received_dttm"].max()}')
print(f'  Rows: {len(df):,}')

Date range before filtering:
  Min: 2025-09-01 00:00:46
  Max: 2026-08-26 23:56:48
  Rows: 359,542


In [17]:
start_date = '2025-09-01'
end_date = '2026-08-31 23:59:59'

df = df[(df['received_dttm'] >= start_date) & (df['received_dttm'] <= end_date)]

print(f'\nAfter filtering ({start_date} to {end_date}):')
print(f'  Rows: {len(df):,}')
print(f'  Min date: {df["received_dttm"].min()}')
print(f'  Max date: {df["received_dttm"].max()}')


After filtering (2025-09-01 to 2026-08-31 23:59:59):
  Rows: 359,542
  Min date: 2025-09-01 00:00:46
  Max date: 2026-08-26 23:56:48


---
## 9. Export Cleaned Data

In [18]:
df.to_csv(OUTPUT_PATH, index=False)
print(f'Exported to: {OUTPUT_PATH}')
print(f'Shape: {df.shape[0]:,} rows x {df.shape[1]} columns')

Exported to: C:\Users\LOQ\OneDrive\Desktop\Emergancy\sf-ems-project\data\cleaned\sf_ems_clean_9_2025_to_8_2026.csv
Shape: 359,542 rows x 56 columns


---
## 10. Final Summary

In [19]:
print('=== FINAL DATASET SUMMARY ===')
print(f'Rows: {len(df):,}')
print(f'Columns: {df.shape[1]}')
print(f'Date range: {df["received_dttm"].min()} to {df["received_dttm"].max()}')
print()

print('Call Type Group distribution:')
print(df['call_type_group'].value_counts().to_string())
print()

print('ALS Unit distribution:')
print(df['als_unit'].value_counts().to_string())
print()

print('Top 10 neighborhoods:')
print(df['neighborhoods_analysis_boundaries'].value_counts().head(10).to_string())
print()

print('Response time stats (minutes, REAL values, not capped):')
print(f'  Count: {df["response_time_min"].notna().sum():,}')
print(f'  Mean: {df["response_time_min"].mean():.2f}')
print(f'  Median: {df["response_time_min"].median():.2f}')
print(f'  Std: {df["response_time_min"].std():.2f}')
print(f'  Min: {df["response_time_min"].min():.2f}')
print(f'  Max: {df["response_time_min"].max():.2f}')
print()

print('Outlier flags (True = flagged):')
for col in iqr_cols:
    print(f'  is_outlier_{col}: {df["is_outlier_" + col].sum():,}')

=== FINAL DATASET SUMMARY ===
Rows: 359,542
Columns: 56
Date range: 2025-09-01 00:00:46 to 2026-08-26 23:56:48

Call Type Group distribution:
call_type_group
Potentially Life-Threatening    177429
Alarm                            90356
Non Life-threatening             69604
Fire                             14200
Unknown                           7953

ALS Unit distribution:
als_unit
True     193562
False    165980

Top 10 neighborhoods:
neighborhoods_analysis_boundaries
Tenderloin                        49101
South of Market                   40533
Mission                           36539
Financial District/South Beach    22003
Bayview Hunters Point             18462
Nob Hill                          13841
Sunset/Parkside                   12853
Western Addition                  12535
Hayes Valley                      10199
Outer Richmond                     8667

Response time stats (minutes, REAL values, not capped):
  Count: 281,553
  Mean: 11.45
  Median: 8.22
  Std: 15.82
  Min: -1

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359542 entries, 0 to 359541
Data columns (total 56 columns):
 #   Column                             Non-Null Count   Dtype         
---  ------                             --------------   -----         
 0   call_number                        359542 non-null  int64         
 1   unit_id                            359542 non-null  object        
 2   incident_number                    359542 non-null  int64         
 3   call_type                          359542 non-null  object        
 4   call_date                          359542 non-null  object        
 5   watch_date                         359542 non-null  object        
 6   received_dttm                      359542 non-null  datetime64[ns]
 7   entry_dttm                         359542 non-null  datetime64[ns]
 8   dispatch_dttm                      359542 non-null  datetime64[ns]
 9   response_dttm                      349916 non-null  datetime64[ns]
 10  on_scene_dttm       

---

## Columns in Cleaned Dataset

| Column | Type | Description |
|--------|------|-------------|
| `call_number` | int64 | Unique 911 call ID |
| `unit_id` | str | Unit dispatched |
| `incident_number` | int64 | Groups units per call |
| `call_type` | str | Specific call type |
| `call_date` | datetime | Call date only |
| `watch_date` | datetime | Shift date |
| `received_dttm` | datetime | 911 call received time |
| `entry_dttm` | datetime | Entered into CAD |
| `dispatch_dttm` | datetime | Unit dispatched |
| `response_dttm` | datetime | Unit acknowledged |
| `on_scene_dttm` | datetime | Unit arrived on scene |
| `transport_dttm` | datetime | Patient transported |
| `hospital_dttm` | datetime | Arrived at hospital |
| `call_final_disposition` | str | Outcome |
| `available_dttm` | datetime | Unit available again |
| `address` | str | Block-level address |
| `city` | str | City name |
| `zipcode_of_incident` | float64 | ZIP code |
| `battalion` | str | Fire battalion |
| `station_area` | float64 | Station zone |
| `box` | str | Dispatch box |
| `original_priority` | str | Initial priority |
| `priority` | str | Updated priority |
| `final_priority` | int64 | Final priority (2 or 3) |
| `als_unit` | bool | ALS unit dispatched |
| `call_type_group` | str | Broad call category |
| `number_of_alarms` | int64 | Alarm level |
| `unit_type` | str | Type of unit |
| `unit_sequence_in_call_dispatch` | int64 | Dispatch order |
| `fire_prevention_district` | float64 | Fire prevention zone |
| `supervisor_district` | float64 | Supervisor district |
| `neighborhoods_analysis_boundaries` | str | Neighborhood name |
| `rowid` | str | Composite key |
| `case_location` | str | GPS coordinates |
| `date` | date | **WEB**: calendar date key |
| `avg_temperature` | float64 | **WEB**: Open-Meteo daily mean deg C |
| `min_temperature` | float64 | **WEB**: Open-Meteo daily min deg C |
| `max_temperature` | float64 | **WEB**: Open-Meteo daily max deg C |
| `precipitation` | float64 | **WEB**: daily precipitation sum (mm) |
| `max_wind_speed` | float64 | **WEB**: daily max wind speed (km/h) |
| `holiday_name` | str | **WEB**: holiday name (NaN if none) |
| `is_holiday` | bool | **WEB**: True if a US/CA holiday |
| `response_time_min` | float64 | **NEW**: On scene - received (min, real values) |
| `dispatch_lag_min` | float64 | **NEW**: Dispatch - entry (min) |
| `transport_time_min` | float64 | **NEW**: Hospital - transport (min, only transports) |
| `hour_of_day` | int64 | **NEW**: Hour (0-23) |
| `day_of_week` | str | **NEW**: Day name |
| `month` | int64 | **NEW**: Month (1-12) |
| `year` | int64 | **NEW**: Year |
| `is_weekend` | bool | **NEW**: Saturday/Sunday |
| `is_als` | int64 | **NEW**: ALS unit (0/1) |
| `is_outlier_response_time_min` | bool | **NEW**: IQR flag (values NOT capped) |
| `is_outlier_dispatch_lag_min` | bool | **NEW**: IQR flag (values NOT capped) |
| `is_outlier_transport_time_min` | bool | **NEW**: IQR flag (values NOT capped) |

---

**Next notebook**: `02_eda.ipynb` - Exploratory Data Analysis